docs/mnist_example/README.md

```diff
@@ -106,6 +106,56 @@ onnx-mlir -O3 --EmitJNI mnist.onnx
 
 A `mnist.jar` should appear, which corresponds to the compiled model object file along with Java API classes.
 
+## Multi-threading
+
+onnx-mlir provides a multi-thread safe parallel compilation mode. Whether each thread is given a name or not by the user, onnx-mlir is multi-threaded safe. If you need to give a name to a thread, use the `-customEnvFlags` keyword and an example can be found as follows.
+
+```bash
+onnx-mlir -O3 -customEnvFlags mnistwith03 [--EmitLib] mnist.onnx -o mnist03
+```
+
+A multi-threaded Python experiment is provided.
+
+```python
+import datetime
+import os
+import threading
+ 
+def execCmd(cmd):
```

Comment:
I guess that works here, aside of the security risk of letting the user choose any exec. I would hardwire onnx-mlir and/or check that the command start with onnx-mlir.

While that approach works, I was suggesting that the system commands occurs in `omCompileFromFile`.

In fact, I am advocating that we deprecate/remove all of the 
```C++
ONNX_MLIR_EXPORT int64_t omSetCompilerOptionsFromEnv(const char *envVarName);
ONNX_MLIR_EXPORT int64_t omSetCompilerOptionsFromArgs(
    int64_t argc, char *argv[]);
ONNX_MLIR_EXPORT int64_t omSetCompilerOptionsFromArgsAndEnv(
    int64_t argc, char *argv[], const char *envVarName);
ONNX_MLIR_EXPORT int64_t omSetCompilerOption(
    const OptionKind kind, const char *val);
ONNX_MLIR_EXPORT void omClearCompilerOption(const OptionKind kind);
ONNX_MLIR_EXPORT void omClearCompilerOption(const OptionKind kind);
```

The new c interface would then be: 
```C++
ONNX_MLIR_EXPORT int64_t omCompileFromFile(const char *inputFilename,
    const char *outputBaseName, EmissionTargetType emissionTarget,
    const char *flags,
    const char **outputFilename, const char **errorMessage);
```

and inside you can simply set a custom env variable (unique name) to the passed flags, and then launch "onnx-mlir --custom-env-var=<uniquename> <input file> -o <output file>"

The preferred way to to use something like this (this one is for llc, but you can do the same for onnx-mlir
```C++
  Command llvmToObj(/*exePath=*/!llcPath.empty() ? llcPath : kLlcPath);
  int rc = llvmToObj.appendStr(getOptimizationLevelOption())
               .appendStr(getTargetTripleOption())
               .appendStr(getTargetArchOption())
               .appendStr(getTargetCPUOption())
               .appendList(getXllcOption())
               .appendStr(getLLVMOption())
               .appendStr("-filetype=obj")
               .appendStr("-relocation-model=pic")
               .appendList({"-o", modelObjNameWithExt})
               .appendStr(bitcodeNameWithExt)
               .exec();
  return rc != 0 ? CompilerFailureInLLVMToObj : CompilerSuccess;
```

Once you have this function, you can change how the python works (and a few example in the code).

If you want help/me to do this infrastructure work,  I can do it.

Reply:
I think I get it. Let me first have a try on this. Thank you.